### PR TITLE
`fix` closed `CacheAdapter`

### PR DESCRIPTION
**Pull request by zilto**
The `CacheAdapter()` would successfully run for the first execution, and fail on subsequent ones. This is because the cache was opened on `__init__()` (needs to read some attributes at instantiation) and closed "after graph execution".

## Changes
cache is now opened and closed at `__init__()`
cache is opened "before graph execution" and closed "after graph execution" for satefy

- [ ] PR has an informative and human-readable title (this will be pulled into the release notes)
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future TODOs are captured in comments
- [ ] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

File: hamilton/lifecycle/default.py [hunk 1]
```diff
@@ -348,9 +348,11 @@ def __init__(
             key=CacheAdapter.nodes_history_key, default=dict()
         )
         self.used_nodes_hash: Dict[str, str] = dict()
+        self.cache.close()
 
     def run_before_graph_execution(self, *, graph: HamiltonGraph, **kwargs):
         """Set `cache_vars` to all nodes if received None during `__init__`"""
+        self.cache = shelve.open(self.cache_path)
```
Comment (elijahbenizzy) [inline hamilton/lifecycle/default.py:355]: Something feels wrong about this. Aren't we just opening it again? Feels like it should:
1. Open during `init`
2. Run during execution
3. Close at the end

**Review by elijahbenizzy — COMMENTED** [review]
Not 100% sure this is the right solution...

File: hamilton/lifecycle/default.py (hunk 1 above)
Comment (zilto) [reply hamilton/lifecycle/default.py:355]: well, that was the previous behavior... I want to have `.close()` *somewhere* for safety, but putting it in `run_after_graph_execution` means you can only run it once

Comment (skrawcz) [reply hamilton/lifecycle/default.py:355]: I think that's what Thierry just did here?

Comment (elijahbenizzy) [reply hamilton/lifecycle/default.py:355]: Ahh yes, my bad

**Review by elijahbenizzy — APPROVED** [review]
Ship it! Sorry misunderstood a bit. This looks good.

**Pull request merged**